T6353: Add cracklib dependencies, fix local Docker build

oniko94 opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change summary
<!--- Provide a general summary of your changes in the Title above -->
- Add the following packages required by vyos/vyos-1x:
  - `libcrack2`
  - `cracklib-runtime`
  - `python3-cracklib`
  - `miscfiles` - Dictionaries
- Fix a bug in Dockerfile that caused local Docker builds always to fail
## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [x] Other (please describe):
  - Broken local build fix: when running `docker build -t vyos/vyos-build:current docker` in the source root directory locally the build was continuously failing on step 12 (line 138) when trying to download and install opam and OCaml, always displaying the same error message. A closer investigation revealed that the line causing the error was `sed -i 's/read -r BINDIR/BINDIR=\"\"/' /tmp/opam_install.sh` which was supposed to edit an installation script on the fly. However, since the installation script is always pulled from the source repository, this line was no longer valid since the current version of the script uses `read_tty` instead of `read -r` (line 1199 of the opam install script). After changing the line in question a local container was successfully built.

  - Add the system dependencies required by password strength check functionality in vyos/vyos-1x

### Local build error message (pre-change):

  ```
  => ERROR [12/45] RUN curl https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh       --output /tmp/opam_install.sh --retry 10 --retry-delay 5 &&     sed -i 's/read   2.1s 
  ------                                                                                                                                                                                      
   > [12/45] RUN curl https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh       --output /tmp/opam_install.sh --retry 10 --retry-delay 5 &&     sed -i 's/read -r BINDIR/BINDIR=""/' /tmp/opam_install.sh && sh /tmp/opam_install.sh &&     opam init --root=/opt/opam --comp=4.14.2 --disable-sandboxing --no-setup:                                                    
  0.173   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                                                                                       
  0.173                                  Dload  Upload   Total   Spent    Left  Speed                                                                                                         
  100 61474  100 61474    0     0   167k      0 --:--:-- --:--:-- --:--:--  167k                                                                                                              
  0.540 ## Downloading opam 2.3.0 for linux on x86_64...
  2.017 ## Downloaded.
  2.017 ## Where should it be installed ? [/usr/local/bin] 
  ------
  Dockerfile:138
  --------------------
   137 |     # Installing OCAML needed to compile libvyosconfig
   138 | >>> RUN curl https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh \
   139 | >>>       --output /tmp/opam_install.sh --retry 10 --retry-delay 5 && \
   140 | >>>     sed -i 's/read -r BINDIR/BINDIR=""/' /tmp/opam_install.sh && sh /tmp/opam_install.sh && \
   141 | >>>     opam init --root=/opt/opam --comp=${OCAML_VERSION} --disable-sandboxing --no-setup
   142 |     
  --------------------
  ERROR: failed to solve: process "/bin/sh -c curl https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh       --output /tmp/opam_install.sh --retry 10 --retry-delay 5 &&     sed -i 's/read -r BINDIR/BINDIR=\"\"/' /tmp/opam_install.sh && sh /tmp/opam_install.sh &&     opam init --root=/opt/opam --comp=${OCAML_VERSION} --disable-sandboxing --no-setup" did not complete successfully: exit code: 1
  ```

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
* https://vyos.dev/T6353

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->
https://github.com/vyos/vyos-1x/pull/4338 - Is blocked by this PR

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

sarthurdev commented:
If these packages are needed at runtime, you should them to the `debian/control` file in vyos-1x. Docker packages are only present during build.

oniko94 commented:
> If these packages are needed at runtime, you should them to the `debian/control` file in vyos-1x. Docker packages are only present during build.

@sarthurdev thank you for the tip!

I've added them to the 'debian/control' in vyos-1x as well, but I'm facing an issue - the 'cracklib-runtime' package should create a binary file `/var/cache/cracklib/cracklib_dict.pwd` which contains the database the Python binding is comparing our passwords against - yet, once I boot the live iso, this file is not there (even though the cracklib binaries themselves, as well as config files in `/etc` and a daily cronjob - are present). I have added the following line at the bottom of `vyos-1x.postinst` script:

```
CRACKLIB_DIR=/var/cache/cracklib
CRACKLIB_DB=cracklib_dict
if [ ! -f "${CRACKLIB_DIR}/${CRACKLIB_DB}.pwd" ]; then
    mkdir -p $CRACKLIB_DIR
    /usr/sbin/create-cracklib-dict -o $CRACKLIB_DIR/$CRACKLIB_DB \
        /usr/share/dict/cracklib-small
fi
```

assuming the cache folder is not being copied or is purged during the installation, but unfortunately they have no effect. At the moment I have to run it manually on logon. If you could share with me a tip in which general direction I should look to resolve this, I will be immensely grateful!

sarthurdev commented:
https://github.com/vyos/vyos-build/blob/current/data/live-build-config/rootfs/excludes#L47

Offending line that is clearing out the caches.

oniko94 commented:
@sarthurdev thank you, I figured it out!

Trying to build vyos-1x without adding these libraries to the Dockerfile still fails, though. I'm not sure whether it should, but at least as far as I get it we need those libraries both at the build stage and during the runtime, so I'll leave this PR up until further notice.

dmbaturin commented on docker/Dockerfile:
```diff
@@ -264,7 +264,11 @@ RUN pip install --break-system-packages \
+      miscfiles \
+      libcrack2 \
+      cracklib-runtime \
+      python3-cracklib
```
There is no need to specify anything but `python3-cracklib` because everything else will be pull as its transitive dependencies anyway. See https://packages.debian.org/bookworm/python3-cracklib to verify,

oniko94 replied:
Removing all of these but `python3-cracklib` results in image build failure, throwing the following exception

```
# lines omitted ...
Del udp-broadcast-relay 0.1+vyos4+current1 [6348 B]
Del ipcalc 0.42-2 [27.8 kB]
Del binutils-common 2.40-2 [2487 kB]
Del kea-admin 2.4.1-3 [444 kB]
Del ndisc6 1.0.5-1+b2 [97.7 kB]
Del telnet 0.17+2.4-2+deb12u1 [41.1 kB]
Del libyaml-0-2 0.2.5-1 [53.6 kB]
Del libwsutil14 4.0.17-0+deb12u1 [100 kB]
Del libfdisk1 2.38.1-5+deb12u3 [194 kB]
Del wireshark-common 4.0.17-0+deb12u1 [484 kB]
Del libsub-uplevel-perl 0.2800-3 [14.0 kB]
Del linux-image-6.6.69-vyos 6.6.69-1 [29.5 MB]
Del python3-idna 3.3-1+deb12u1 [41.0 kB]
Del cryptsetup-bin 2:2.6.1-4~deb12u2 [474 kB]
Del udev 252.33-1~deb12u1 [1704 kB]
Del libxmlsec1 1.2.37-2 [153 kB]
Del libtirpc3 1.3.3+ds-1 [85.2 kB]
Del gettext-base 0.21-12 [160 kB]
Traceback (most recent call last):
  File "/vyos/./build-vyos-image", line 681, in <module>
    cmd("lb build 2>&1")
  File "/vyos/scripts/image-build/utils.py", line 84, in cmd
    raise OSError(f"Command '{command}' failed")
OSError: Command 'lb build 2>&1' failed
```
